src/renderers/WebGLRenderer.js

```diff
@@ -1502,6 +1503,10 @@ function WebGLRenderer( parameters = {} ) {
 
 				needsProgramChange = true;
 
+			} else if ( materialProperties.toneMapping !== _this.toneMapping ) {
+
+				needsProgramChange = true;
```

Comment:
Here's the change I made to force shaders to recompile if the tone mapping changes, which seems like the right thing to do? This was how I fixed the bug with the Sky example.

Reply:
> Here's the change I made to force shaders to recompile if the tone mapping changes, which seems like the right thing to do?

Yes, that is correct.

I think this issue did not happen earlier since it's probably a special case to render a scene with different tone mapping settings.

Reply:
>Yes, that is correct.

Are you sure? I mentioned below that you need to consider `material.toneMapped`.  Right?

> it's probably a special case to render a scene with different tone mapping settings.

PMREM does it. CubeCamera should, for proper workflow. RenderTargets can do it.

Reply:
IIUC, `material.toneMapped` is already accounted for, as it is used in generating the  parameters that go into the cacheKey. The trouble is that three has two levels of  deciding to recompile a material: First `needsProgramChange` has to be true in the renderer in order to even calculate a new cacheKey, then the cacheKey  has to be unique to recompile. I don't really understand why we have both; it seems likely to cause  bugs like this. I assume it's a performance optimization?

Reply:
Ah, I see what you mean. When `Material.toneMapped` is set to `false`, `NoToneMapping` for the program parameters. That means it should be:
```js
const toneMapping = material.toneMapped ? _this.toneMapping : NoToneMapping;
```
and further down:
````
} else if ( materialProperties.toneMapping !== toneMapping ) {
```

Reply:
I've updated the branch with this fix: 04d841a7cd913ee3b0780880b6d7b6d3138bb33b

Reply:
> I don't really understand why we have both; it seems likely to cause bugs like this. I assume it's a performance optimization?

In the earlier days, there was a 1:1 mapping of materials and shader programs. Over time this assumption was no longer true. A single material can produce different shaders depending on which 3D object it uses (e.g. mesh vs instanced mesh), how the renderer or the scene is configured.

The mentioned code section in `setProgram()` exists to automatically detect these use cases. Without it, it would be necessary to manually trigger recompiles via setting `Material.needsIUpdate` to `true` OR by computing the program cache key per frame.

Reply:
Maybe we can optimize the program cache key computation to a point where we can execute it per frame. That would simplify the code. However, there is #22530.

Reply:
Ah, I see. I *think* the behavior is the same, but it'll skip some cacheKey checks. SG. Should I pull your change into my PR here?

Reply:
Oh, never mind you already updated it; thanks!